### PR TITLE
Now working.

### PR DESCRIPTION
**Pull request by athornton**

### CODE REVIEW

File: manifests/base/configmap-m.yaml [hunk 1]
```diff
@@ -0,0 +1,18 @@
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: m-cm
```
Comment (rra) [inline manifests/base/configmap-m.yaml:4]: Can you add the word moneypenny to the name? It helps when looking at all the resources in the environment

File: manifests/base/configmap-quips.yaml [hunk 2]
```diff
@@ -0,0 +1,7 @@
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: quips-cm
```
Comment (rra) [inline manifests/base/configmap-quips.yaml:4]: Likewise here

File: manifests/base/deployment.yaml [hunk 3]
```diff
@@ -1,27 +1,43 @@
...
   selector:
     matchLabels:
-      name: cachemachine
+      name: moneypenny
   template:
     metadata:
       labels:
-        name: cachemachine
+        name: moneypenny
     spec:
```
Comment (rra) [inline manifests/base/deployment.yaml:16]: If this is the live manifest that we'll be using in production, it should have the pod and container hardening settings from SQR-048 (except for disabling mounting the service account token, of course).

File: scripts/start.sh [hunk 4]
```diff
@@ -1,4 +1,4 @@
 #!/bin/bash -ex
 
-curl --insecure -X POST -H "Content-Type: application/json" -d "@jupyter.json" https://minikube.lsst.codes/cachemachine/
-curl --insecure -X POST -H "Content-Type: application/json" -d "@simple.json" https://minikube.lsst.codes/cachemachine/
+curl --insecure -X POST -H "Content-Type: application/json" -d "@jupyter.json" https://minikube.lsst.codes/moneypenny/
+curl --insecure -X POST -H "Content-Type: application/json" -d "@simple.json" https://minikube.lsst.codes/moneypenny/
```
Comment (rra) [inline scripts/start.sh:4]: I think you wanted to just delete this file. It doesn't seem relevant to you.

File: src/moneypenny/config.py [hunk 5]
```diff
@@ -40,3 +40,42 @@ class Configuration:
...
+    at default in normal operation.
+    """
+
+    dossier_path: str = os.getenv("DOSSIER_FILE", "/opt/dossier")
+    """Path to the dossier file's mountpoint created in the initContainers.
+    Leave at default in normal operation.  Note that it does not contain the
+    filename (usually "dossier.json")
+    """
+
+    M: str = os.getenv("M_FILE", f"/{config_dir}/M/m.yaml")
```
Comment (rra) [inline src/moneypenny/config.py:57]: Nit: Single-letter all-caps variable names are normally reserved for type variables. Maybe `m_config`? Also, `M_CONFIG` seems like a better name here, since this is a configuration file specifically.

File: src/moneypenny/config.py [hunk 6]
```diff
@@ -40,3 +40,42 @@ class Configuration:
...
+    """Path to M's standing orders.  Leave at default in normal operation.
+    """
+
+    quips: str = os.getenv("QUIP_FILE", f"{config_dir}/quips/quips.txt")
```
Comment (rra) [inline src/moneypenny/config.py:61]: Nit: Environment variable needs a `MONEYPENNY_` prefix.

File: src/moneypenny/config.py [hunk 7]
```diff
@@ -40,3 +40,42 @@ class Configuration:
...
+    """Path to Moneypenny's quip file.  Leave at default in normal operation.
+    """
+
+    moneypenny_timeout: int = int(os.getenv("MONEYPENNY_TIMEOUT") or "300")
+    """Timeout (in seconds) to wait for all containers in the action pod to
+    complete.  Defaults to 300.
+    """
+
+    null_container_image: str = os.getenv(
+        "NULL_CONTAINER_IMAGE", "lsstsqre/farthing:latest"
```
Comment (rra) [inline src/moneypenny/config.py:71]: I'm confused by why you need this. My understanding is that the goal is to run a stack of containers one after the other. Why not just take the last container and run it as the main container and the rest as init containers? Seems equivalent and doesn't require this dummy container.

File: src/moneypenny/config.py [hunk 8]
```diff
@@ -40,3 +40,42 @@ class Configuration:
...
+    )
+    """Docker image specification for an "exit with return code zero"
+    do-nothing image.  Defaults to "lsstsqre/farthing:latest" but, for
+    instance, "library/alpine:latest" works fine too.
+    """
+
+    # This is not user-configurable; K8s specifies it.
+    namespace_file: str = (
+        "/var/run/secrets/kubernetes.io/serviceaccount/namespace"
+    )
```
Comment (rra) [inline src/moneypenny/config.py:81]: Nit: If it's not configurable, I would move it into a `moneypenny.constants` module.

File: src/moneypenny/errors.py [hunk 9]
```diff
@@ -0,0 +1,45 @@
+"""Helper globals and errors for Moneypenny.
+"""
+
+namespace_file = "/var/run/secrets/kubernetes.io/serviceaccount/namespace"
+primary_container_image = "library/alpine:latest"
+dossier_file = "/opt/lsst/software/moneypenny/config/dossier/dossier.yaml"
+M = "/opt/lsst/software/moneypenny/config/M/m.yaml"
+quips = "/opt/lsst/software/moneypenny/config/quips/quips.txt"
+moneypenny_timeout = 300
```
Comment (rra) [inline src/moneypenny/errors.py:9]: These seem to reiterate the constants from the config? Not sure why they're here.

Nit: For other projects, we've called the module collecting the exceptions `moneypenny.exceptions`.

File: src/moneypenny/handlers/external/agent.py [hunk 10]
```diff
@@ -36,23 +37,26 @@ async def check_status(request: web.Request) -> web.Response:
      a 202 response if it is still in progress, and raises an error otherwise.
     If the pod cannot be found, a 404 is raised; otherwise a 500 with
      descriptive text is raised.
+
+    Because Moneypenny tidies up rather quickly, a 200 is very unlikely,
+     a 202 fairly unlikely, and a 404 common.
```
Comment (rra) [inline src/moneypenny/handlers/external/agent.py:42]: Given this, is there much point in this route? I'm not sure when I would use it or why.

File: src/moneypenny/handlers/external/agent.py [hunk 11]
```diff
@@ -68,20 +72,21 @@ async def commission_agent(request: web.Request) -> web.Response:
...
 
 
-
 @routes.delete("/{username}")
 async def retire_agent(request: web.Request) -> web.Response:
     """DELETE /moneypenny/{username}
 
     Perform deprovisioning steps for the given username.  Returns a 202 once
     the request has been accepted.
     """
```
Comment (rra) [inline src/moneypenny/handlers/external/agent.py:86]: If you're going to have a DELETE, the GET on `/{username}` needs to return success for all provisioned users, not just the ones with pods running, or the semantics get screwy (you can successfully DELETE a resource that returns 404).

I don't think DELETE semantics are what you want, since Moneypenny has no idea what users exist. Unless you're going to add knowledge to Moneypenny of what users exist (which doesn't seem right), I think deprovisioning should be a POST to some other route. It's a command, not a resource management instruction, so REST says that should be a POST. That may argue for not putting the POST for provisioning at the top level either; my inclination would be to have `/provision` and `/deprovision` routes and not try to be all RESTful here since REST semantics don't really fit (you're not managing a resource collection).

File: src/moneypenny/handlers/external/agent.py [hunk 12]
```diff
@@ -68,20 +72,21 @@ async def commission_agent(request: web.Request) -> web.Response:
         ),
     )
     moneypenny = request.config_dict["moneypenny"]
-    username = body["username"]
+    username = body["token"]["data"]["uid"]
     await moneypenny.execute_order(verb="post", dossier=body)
```
Comment (rra) [inline src/moneypenny/handlers/external/agent.py:76]: When you're passing in the name of the function to call as a string argument to a method, it feels like something has gone wrong. Why isn't the internal API just `provision()` and `deprovision()` methods?

File: src/moneypenny/kubernetes.py [hunk 13]
```diff
@@ -1,165 +1,315 @@
...
+    All Exceptions raised are our own, although they may well be thin
+    wrappers around the corresponding K8s API error.
     """
 
-    def __init__(self) -> None:
-        """Create a new client for the cluster we are running in."""
-        self.namespace = Path(NAMESPACE_FILE).read_text().strip()
+    def __init__(self, moneypenny) -> None:  # type: ignore
+        """Create a new client for the cluster we are running in.
+        We ignore type in order to prevent a circular dependency."""
```
Comment (rra) [inline src/moneypenny/kubernetes.py:54]: You're only passing in the moneypenny object to get the config. Just pass in the config directly and that avoids the circular dependency. The Kubernetes client should never need to know about the moneypenny object; that would be a layering violation.

File: src/moneypenny/kubernetes.py [hunk 14]
```diff
@@ -1,165 +1,315 @@
...
+            )
+        self.moneypenny = moneypenny
+        self.config: Configuration = moneypenny.config
+        namespace = "default"
+        nsf = self.config.namespace_file
+        try:
+            namespace = Path(nsf).read_text().strip()
+        except FileNotFoundError:
+            logger.warn(f"Namespace file {nsf} not found; using 'default'")
+        self.namespace = namespace
```
Comment (rra) [inline src/moneypenny/kubernetes.py:67]: This feels weird; why not load the namespace when the config is initialized? It doesn't feel like the Kubernetes client should need to know this implementation detail of how the configuration is created.

File: src/moneypenny/schemas/post.json [hunk 15]
```diff
@@ -7,46 +7,99 @@
...
-	  "name": "doubleos",
-	  "id": 500
-	},
-	{
-	  "name": "staff",
-	  "id": 200
-	}
-      ]
+      "token": {
+        "data": {
```
Comment (rra) [inline src/moneypenny/schemas/post.json:11]: This is all set up to match the legacy token format, which in turn matches the NCSA-adulterated CILogon tokens. All of this is going away in the long run, so I'd rather not incorporate this mess into another application. Instead, could you define the parameters using the same names as are used in the new API documented in SQR-049? That will keep everything consistent in the future and will make it much simpler and also avoid some of the weirdness of the legacy token schema, at the cost of currently needing to convert `/auth/analyze` results to the new format (which will go away once the new Gafaelfawr is deployed).

The schema would therefore look something more like this (probably without the name since you don't care):
```
{
  "username": "alice",
  "name": "Alice Example",
  "uid": 24187,
  "groups": [
    {
      "id": 4173,
      "name": "example-group"
    },
    {
      "id": 5671,
      "name": "other-group"
    }
  ]
}
```
(example from the `/auth/api/v1/user-info` route).

File: src/moneypenny/singleton.py [hunk 16]
```diff
@@ -1,15 +1,16 @@
...
     """
 
-    _instances = {}
+    _instances: Dict[Any, Any] = {}
 
-    def __call__(cls, *args, **kwargs):
+    def __call__(cls, *args, **kwargs) -> Dict[Any, Any]:  # type: ignore
         if cls not in cls._instances:
             cls._instances[cls] = super(Singleton, cls).__call__(
                 *args, **kwargs
```
Comment (rra) [inline src/moneypenny/singleton.py:16]: Do you really need all this? It seems like extra complexity and I'm not sure what it's gaining you.

File: tests/conftest.py [hunk 17]
```diff
@@ -228,134 +235,90 @@ def create_resource(kube_client, kube_ns, resource_type, manifest, delete_first=
...
-@pytest.fixture
-def hub_ssl(kube_ns, hub_pod_ssl):
-    """Return the Hub object for connecting to a running hub pod with internal_ssl enabled"""
-    return Hub(
-        proto="https",
-        ip=f"{hub_pod_ssl.metadata.name}.{kube_ns}",
-        port=8081,
-        base_url="/hub/",
-    )
+# Base create_moneypenny_pod on this et seq
```
Comment (rra) [inline tests/conftest.py:238]: Reminder to delete all the commented-out code before merging.

File: tests/conftest.py [hunk 18]
```diff
@@ -1,37 +1,32 @@
 """pytest fixtures for moneypenny"""
```
Comment (rra) [inline tests/conftest.py:1]: Aie, we clearly need to move all this stuff into a library rather than duplicating this much code or having this giant fixtures configuration file. But we can tackle that later.

**Review by rra — COMMENTED** [review]
Are we deploying this with kustomize and not with Helm? I'm fine if so, but wanted to check since normally we use Helm for RSP services.

Nit: I'm not a fan of relative imports, and they create some minor maintenance problems if you rename modules. I'd rather use absolute imports everywhere.

File: tests/conftest.py [hunk 19]
```diff
@@ -228,134 +235,90 @@ def create_resource(kube_client, kube_ns, resource_type, manifest, delete_first=
...
+#         )
+#         volume_mounts.append(
+#             {
+#                 "mountPath": "/etc/jupyterhub/secret",
+#                 "name": "secret",
+#             }
+#         )
+
+#     pod_manifest = V1Pod(
+#         metadata={
```
Comment (cbanek) [inline tests/conftest.py:278]: Yeah this test file has me confused?  This looks like a bunch of fixtures, which seem reasonable, but the mentions of hub and jupyter confuse me.  And I'm not sure which of these functions are the actual tests?  Can we move those to a different python file in this directory?

File: src/moneypenny/errors.py [hunk 20]
```diff
@@ -0,0 +1,45 @@
+"""Helper globals and errors for Moneypenny.
+"""
+
+namespace_file = "/var/run/secrets/kubernetes.io/serviceaccount/namespace"
```
Comment (cbanek) [inline src/moneypenny/errors.py:4]: Probably don't need the namespace file since it's in the configuration object?

File: src/moneypenny/kubernetes.py (hunk 13 above)
Comment (cbanek) [reply src/moneypenny/kubernetes.py:54]: Yes completely agree with this.  I don't even think you need to pass in the config I think you could make a new Configuration class here and just load it right from the environment variables.  Sure it loads the environment variables a couple of times, but they aren't going to change.

File: src/moneypenny/config.py (hunk 8 above)
Comment (cbanek) [reply src/moneypenny/config.py:81]: Actually, since it's not configurable, and nobody cares about what the path is, but rather what the namespace is, you could make this a property that reads the file, or just read the file at __init__ time and store it as a property here.  That would also clean up some of the namespace parts in the kubernetes.py also reading the namespace file

File: src/moneypenny/kubernetes.py [hunk 21]
```diff
@@ -1,165 +1,315 @@
...
-        phase = status.phase
+        phase: str = status.phase
         if phase == "Succeeded":
             return True
         if phase == "Pending" or "Running":
             return False
         if phase == "Unknown":
-            raise OperationFailed(f"Pod {name} in Unknown phase")
+            raise OperationFailed(f"Pod {pname} in Unknown phase")
         raise OperationFailed(status.message)
```
Comment (cbanek) [inline src/moneypenny/kubernetes.py:174]: This error seems a bit weird since it seems to be a catchall after 173?

File: src/moneypenny/kubernetes.py [hunk 22]
```diff
@@ -1,165 +1,315 @@
...
-        
+
+    def _make_pod(
```
Comment (cbanek) [inline src/moneypenny/kubernetes.py:176]: Okay, after looking at this, and thinking about it a bit, I'm not sure I like this pattern.  First, there's a lot of functions to build it up, but I think they all just call each other once, so it's kind of building up the data, but it's more tricky for me to think about.  Maybe it is better to just have this as a yaml file that is templated.  Because what if we need another type, like a secret or a different configmap?  You can checkout the nublado2 part about just making objects from arbitrary yaml.  We could just have a convention that says there's a pod named "foo" and we wait for it to finish, using the code you've got here to check

Comment (cbanek) [reply src/moneypenny/kubernetes.py:176]: This might mean that we want to isolate these things in different namespaces though maybe... just to be able to delete everything in one go by cleaning up the namespace, so that might be a tricky change.  With this you have to kind of keep track of the cases where things exist.

File: src/moneypenny/moneypenny.py [hunk 23]
```diff
@@ -5,32 +5,30 @@
...
-from .types import CatGotYourTongueError, NonsensicalOrderError
-
-# These should be supplied as ConfigMaps
-M="/opt/lsst/software/moneypenny/config/m.yaml"
-QUIPS="/opt/lsst/software/moneypenny/config/quips.txt"
 
 logger = structlog.get_logger(__name__)
 
 
 class Moneypenny(metaclass=Singleton):
```
Comment (cbanek) [inline src/moneypenny/moneypenny.py:30]: Yeah don't do this, there's a whole post about how aiohttp is opinionated against singletons, so that's why there's init thing.  https://docs.aiohttp.org/en/stable/web_advanced.html#data-sharing-aka-no-singletons-please

Either way you don't need to worry about creating too many of those with special class sugar

File: src/moneypenny/singleton.py (hunk 16 above)
Comment (cbanek) [reply src/moneypenny/singleton.py:16]: Yeah dump this file

File: src/moneypenny/moneypenny.py [hunk 24]
```diff
@@ -149,9 +192,8 @@ async def check_completed(self, username: str) -> Bool:
         ApiException if there was some other error
 
         """
```
Comment (cbanek) [inline src/moneypenny/moneypenny.py:194]: This doc string is out of date

File: src/moneypenny/schemas/post.json [hunk 25]
```diff
@@ -7,46 +7,99 @@
...
+	    }
+	  },
+	  "required": [
+	    "uid",
+	    "uidNumber",
+	    "isMemberOf"
+	  ]
+	}
+      },
+      "additionalProperties": true,
```
Comment (cbanek) [inline src/moneypenny/schemas/post.json:95]: trailing whitespace

File: src/moneypenny/handlers/external/agent.py (hunk 11 above)
Comment (cbanek) [reply src/moneypenny/handlers/external/agent.py:86]: Yeah, I think the symantics here are a bit confusing.  First off, passing down verb as far as as the kubernetes layer seems confusing.  It's a terrible word even for HTTP, and worse when we are using it.  We could call it operation, or task or anything like that.  Pick a pun, just not verb.

That being said, you might want to think of this more like a job system, like there's a /provision endpoint where you post the body, and it hands you a URL that is /provision/cbanek-123456 randomnumber or uuid or something.  That way you can check the status after it's finished and hold onto it until the caller calls DELETE on /provision/cbanek-123456 to cleanup.  The resources are actually what's happening, since what if a job gets stuck or takes a long time, etc.  Also, in the future when there may be more than one of these jobs running per user or those symantics come into play it would probably make sense to treat this more as a job system.

At the http layer have each URL be a job like /provision /deprovision /adminjob2 /takeoutback etc.  Then each of those can have things that are going.  You could just hold it in memory.  Then at the moneypenny layer it figures out the yaml or the resources, and passes that into the kubernetes layers?

File: src/moneypenny/moneypenny.py [hunk 26]
```diff
@@ -122,16 +130,51 @@ async def execute_order(self, verb: str, dossier: Dict[str,Any]) -> None:
         dossier: Dossier associated with the order for the user.
 
         """
-        verb=verb.strip().lower()
+        verb = verb.strip().lower()
         # We re-read the orders each time; the config map is allowed to change
         await self._read_orders()
```
Comment (cbanek) [inline src/moneypenny/moneypenny.py:135]: If you're going to re-read this all the time, instead of using a member variable which implies caching, why not just have this return the thing stored in self.orders and get rid of that entirely?  Then it's obvious you aren't using the cached one.  For example, you are also setting this in init and I don't know if that's used before it is reread?

File: src/moneypenny/moneypenny.py [hunk 27]
```diff
@@ -122,16 +130,51 @@ async def execute_order(self, verb: str, dossier: Dict[str,Any]) -> None:
...
+        expiry = datetime.datetime.now() + datetime.timedelta(seconds=tmout)
+        # Wait for order to complete
+        logger.info(f"Awaiting completion for '{verb}': {username}")
+        count = 0
+        while True:
+            count += 1
+            logger.info(f"Checking on {username}: attempt #{count}")
+            finito = await self.check_completed(username)
+            if finito:
+                break
```
Comment (cbanek) [inline src/moneypenny/moneypenny.py:162]: Instead of using a break, can we just do the cleanup here?  Having a while true with a break is not the best

File: build_tag.sh [hunk 28]
```diff
@@ -0,0 +1,12 @@
+#!/bin/sh
+
+set -e
+R="lsstsqre/moneypenny"
+T="dev"
+I="${R}:${T}"
+L="${R}:latest"
+
+docker build -t ${I} . --platform=amd64
```
[the remark this answers is not in the document]
Comment (athornton) [reply build_tag.sh:9]: But that would violate DRY!

File: manifests/base/configmap-quips.yaml (hunk 2 above)
Comment (athornton) [reply manifests/base/configmap-quips.yaml]: done

File: scripts/start.sh (hunk 4 above)
Comment (athornton) [reply scripts/start.sh]: done

File: src/moneypenny/kubernetes.py (hunk 14 above)
Comment (athornton) [reply src/moneypenny/kubernetes.py]: I've moved the namespace file into being a constant in the kubernetes client; it is fixed (but not present if you're running in a no-RBAC no-namespaces cluster)

File: src/moneypenny/singleton.py (hunk 16 above)
Comment (athornton) [reply src/moneypenny/singleton.py:16]: removed

File: manifests/base/configmap-m.yaml [hunk 29]
```diff
@@ -0,0 +1,18 @@
+apiVersion: v1
```
[the remark this answers is not in the document]
Comment (athornton) [reply manifests/base/configmap-m.yaml]: I would prefer Helm.

File: manifests/base/configmap-m.yaml [hunk 30]
```diff
@@ -0,0 +1,18 @@
...
+kind: ConfigMap
+metadata:
+  name: m-cm
+data:
+  m.yaml: |
+    post:
+      - name: hello
+        image: library/hello
+        securityContext:
+          runAsUser: 1000
```
[the remark this answers is not in the document]
Comment (athornton) [reply manifests/base/configmap-m.yaml]: fixed

File: tests/conftest.py (hunk 19 above)
Comment (athornton) [reply tests/conftest.py:278]: This bit isn't done yet.  I'm stealing the harness from KubeSpawner for the actual things-that-do-things-in-k8s tests, and I'll have some tests around in-memory representations for K8s objects not submitted to kubectl.

File: src/moneypenny/errors.py (hunk 20 above)
Comment (athornton) [reply src/moneypenny/errors.py]: done

File: src/moneypenny/kubernetes.py (hunk 21 above)
Comment (athornton) [reply src/moneypenny/kubernetes.py:170]: Clarified.  We reach that only if pod.status == "Failed" in which case we want to return the message as well since it may tell us why the pod failed.

File: src/moneypenny/kubernetes.py (hunk 22 above)
Comment (athornton) [reply src/moneypenny/kubernetes.py:172]: I've made all the containers just use yaml.  I was running this all in the Moneypenny namespace, because I don't think we need to give Moneypenny cluster privs and at least so far the provisioning/deprovisioning is namespace-independent.  There aren't that many things created: one pod and one configmap per user.

File: src/moneypenny/moneypenny.py (hunk 24 above)
Comment (athornton) [reply src/moneypenny/moneypenny.py:187]: fixed

File: src/moneypenny/moneypenny.py (hunk 26 above)
Comment (athornton) [reply src/moneypenny/moneypenny.py]: I got rid of self.orders and self.dossier entirely.

File: src/moneypenny/moneypenny.py (hunk 27 above)
Comment (athornton) [reply src/moneypenny/moneypenny.py]: Fixed.  I also made it not a while true.

File: src/moneypenny/moneypenny.py (hunk 23 above)
Comment (athornton) [reply src/moneypenny/moneypenny.py]: OK, Singleton dumped.

File: manifests/base/role.yaml [hunk 31]
```diff
@@ -0,0 +1,8 @@
+kind: Role
+apiVersion: rbac.authorization.k8s.io/v1
+metadata:
+  name: moneypenny
+rules:
+- apiGroups: [""]
+  resources: [ "pods", "configmaps"]
+  verbs: ["get", "list", "create", "watch", "delete"]
```
[the remark this answers is not in the document]
Comment (athornton) [reply manifests/base/role.yaml:8]: I can break it into two things, but I do need to be able to at least get/create/delete configmaps.

[the remark this answers is not in the document]
Comment (athornton) [reply manifests/base/role.yaml:8]: Well, create/delete anyway.

File: src/moneypenny/schemas/post.json (hunk 25 above)
Comment (athornton) [reply src/moneypenny/schemas/post.json]: fixed

File: tests/conftest.py (hunk 17 above)
Comment (athornton) [reply tests/conftest.py:238]: tests are very far from being ready, but I will

File: src/moneypenny/config.py (hunk 5 above)
Comment (athornton) [reply src/moneypenny/config.py]: fixed

File: build_tag.sh (hunk 28 above)
[the remark this answers is not in the document]
Comment (athornton) [reply build_tag.sh:9]: More seriously I'm probably going to remove this when merging; this is purely for my own convenience to be able to do build/tag/push with up-arrow a few times and Enter.

File: manifests/base/role.yaml (hunk 31 above)
[the remark this answers is not in the document]
Comment (rra) [reply manifests/base/role.yaml]: This is all constrained to the `moneypenny` namespace, so I think it's fine. I would be pickier about the exact permissions if it were a `ClusterRole`, but since all that moneypenny could mess with is its own configmaps, I have no security worries here.

File: manifests/base/deployment.yaml (hunk 3 above)
Comment (athornton) [reply manifests/base/deployment.yaml:16]: done

File: src/moneypenny/schemas/post.json (hunk 15 above)
Comment (athornton) [reply src/moneypenny/schemas/post.json]: Done

File: src/moneypenny/config.py (hunk 8 above)
Comment (athornton) [reply src/moneypenny/config.py]: Removed; it's just a file-scoped constant in the kubernetes driver now.

File: src/moneypenny/errors.py (hunk 9 above)
Comment (athornton) [reply src/moneypenny/errors.py]: constants removed, errors->exceptions

File: src/moneypenny/handlers/external/agent.py (hunk 10 above)
Comment (athornton) [reply src/moneypenny/handlers/external/agent.py:43]: Because we submit the request and don't wait for it to complete.  So if you wanted to write a UI around Moneypenny, for instance, you'd submit a request and then show a spinner while you got a 202, returning "it completed" when you get a 200 or a 404.

File: src/moneypenny/handlers/external/agent.py (hunk 11 above)
Comment (athornton) [reply src/moneypenny/handlers/external/agent.py]: Changed to multiple endpoints, one per action, with a POST.

We actually only want one operation going at a time for any given user.  If a user receives another request while it has one being processed, you get back a 409 Conflict.

File: src/moneypenny/handlers/external/agent.py (hunk 12 above)
Comment (athornton) [reply src/moneypenny/handlers/external/agent.py]: Well, we now have "commission" and "retire" but they're still calling a single function back in moneypenny, because what moneypenny does to execute the function is to construct a Pod with a sequence of containers, and execute them in order, no matter what the function is.

File: src/moneypenny/kubernetes.py (hunk 13 above)
Comment (athornton) [reply src/moneypenny/kubernetes.py]: I'm passing the config down.  I guess I *could* make one there, but...meh.  Six of one...

File: src/moneypenny/config.py (hunk 6 above)
Comment (athornton) [reply src/moneypenny/config.py:61]: done

File: src/moneypenny/config.py (hunk 7 above)
Comment (athornton) [reply src/moneypenny/config.py:71]: removed

File: tests/conftest.py (hunk 17 above)
Comment (athornton) [reply tests/conftest.py]: Done

File: tests/conftest.py (hunk 18 above)
Comment (athornton) [reply tests/conftest.py]: I have a much smaller conftest now.

File: build_tag.sh (hunk 28 above)
[the remark this answers is not in the document]
Comment (athornton) [reply build_tag.sh]: removed

[the remark this answers is not in the document]
Comment (athornton) [reply build_tag.sh]: removed

**Review by rra — APPROVED** [review]
Yeah, this all looks good to me. Thank you! The only thing I'm still a bit confused by is the Kubernetes configs here in this repository, but without the Kustomize wrapper. I would have expected either Kustomize configuration in this repo, or for all the Kubernetes stuff to move to the charts repo.

**Comment by athornton** [comment]
>I would have expected either Kustomize configuration in this repo, or for all the Kubernetes stuff to move to the charts repo.

I'll move it to charts, once I get this merged and it's ready to go into charts, but I don't want to start messing with charts until I have the service ready to be deployed into it.  There's some more work I need to do on rubin_jupyter_hub to start using Moneypenny too.
